### ISSUE TITLE
Tuple through try/except gives object has no attribute __iter__

### ISSUE THREAD

**Issue by scaramallion** [python/mypy#10556]
I'm getting a false positive with:

```python
from typing import Tuple, Optional


t: Tuple[Optional[bool], Optional[bool]]
try:
    t = (True, False)
except:
    t = (None, None)

reveal_type(t)  # 'Tuple[Union[builtins.bool, None], Union[builtins.bool, None], fallback=builtins.object]'
None in t  # "object" has no attribute __iter__
```
While this works as expected:
```python
t: List[Optional[bool]]
try:
    t = [True, False]
except:
    t = [None, None]

None in t
```
**Your Environment**

- Mypy version used: 0.812
- Python version used: 3.9

**Comment by hauntsaninja** [comment]
Hmm, the fallback there is suspicious. It looks like mypy is getting an incorrect result for joining the fallbacks of the tuples in the try-except: https://github.com/python/mypy/blob/61c346230cf4960c9766f7c2b5bd3312b47f74da/mypy/join.py#L285 (maybe it's not taking into account that tuple is covariant?)

**Comment by hauntsaninja** [comment]
Yeah, that looks to be the case. #5269 is the issue for this generally. It looks like @freundTech has already written a PR implementing this — I can confirm that it fixes your issue :-)

**Comment by scaramallion** [comment]
Yep, can confirm that the PR fixes it. Thanks!
